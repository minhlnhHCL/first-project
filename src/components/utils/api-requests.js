import axios from 'axios';

export const baseUrl = 'https://dummyjson.com/users';

export const get = () => {
  return axios.get(baseUrl);
};

export const post = (url, body) => {
  return axios.post(baseUrl + url, body);
};

export const put = (url, body) => {
  return axios.put(baseUrl + url, body);
};

export const deleteReq = (url, body) => {
  return axios.delete(baseUrl + url, body);
};
