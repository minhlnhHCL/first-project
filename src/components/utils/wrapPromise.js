export const wrapPromise = (promise) => {
    // Initial status
    let status = 'pending'

    let result;
    const suspender = promise.then(
        res => {
            status = 'success'
            result = res
        },
        err => {
            status = 'error'
            result = err
        }
    )
    const read = () => {
        switch (status) {
            case 'pending':
                throw suspender
            case 'error':
                throw result
            default:
                return result
        }
    }

    return { read }
}