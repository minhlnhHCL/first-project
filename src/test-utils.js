import { render } from "@testing-library/react"
import { Provider } from "react-redux"
import { BrowserRouter, Routes } from "react-router-dom"
import { SpinnerContext, SpinnerProvider } from "./context/SpinnerContext"
import setupStore from "./store"

// export const renderWithRedux = (children) => {
//     const renderResult = render(<Provider store={store}>
//         <SpinnerProvider value={SpinnerContext}>
//             {children}
//         </SpinnerProvider>
//     </Provider>)
//     return {
//         ...renderResult,
//         store,
//     };
// }
// re-export everything
export * from '@testing-library/react'

export const renderWithRedux = (
    ui,
    {
        preloadedState = {},
        // Automatically create a store instance if no store was passed in
        store = setupStore(preloadedState),
        ...renderOptions
    } = {}
) => {
    const Wrapper = ({ children }) => {
        return (
            <Provider store={store}>
                <SpinnerProvider value={SpinnerContext}>
                    {children}</SpinnerProvider>
            </Provider>
        )
    }
    return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) }
}