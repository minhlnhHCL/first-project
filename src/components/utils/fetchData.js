import axios from "axios"
import { wrapPromise } from "./wrapPromise"

export const fetchData = (url) => {
    const promise = axios.get(url)
        .then(res => res.data)
        .catch(err => console.log(err))

    return wrapPromise(promise)
}