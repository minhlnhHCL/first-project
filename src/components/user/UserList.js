import React, { useEffect, useState } from 'react'
import './UserList.scss'
import UserWrapper from './UserWrapper/UserWrapper'
import { fetchData } from '../utils/fetchData'
import { useNavigate } from 'react-router-dom'
import { useDispatch, useSelector } from 'react-redux'
import { getUsers, setCurrentUser } from '../../store/userSlice'
import { useContext } from 'react'
import { SpinnerContext } from '../../context/SpinnerContext'
import axios from 'axios'

const resource = fetchData('https://dummyjson.com/users')

const UserList = () => {
    const { users } = useSelector(state => state.users)
    const { onLoading, offLoading } = useContext(SpinnerContext)
    const dispatch = useDispatch()
    const data = resource.read()
    const [, setActive] = useState(false)
    const navigate = useNavigate()


    useEffect(() => {
        onLoading()
        const getData = async () => {
            try {
                const res = await axios.get('https://dummyjson.com/users')
                if (res.status == 200) {
                    if (users.length == 0 || users.length == 1) {
                        return dispatch(getUsers(data.users))
                    }
                }
            } catch (error) {
                console.log(error)
            } finally {
                offLoading()
            }
        }
        getData()
        // if (reduxState.users.length == 0 || reduxState.users.length == 1) {
        //     setActive(true)
        //     dispatch(getUsers(data.users))
        // } else {
        //     setActive(false)
        // }
    }, [users])
    return (
        <div className='list-container mt-32'>
            <button className='btn-create' onClick={() => {
                dispatch(setCurrentUser({ firstName: '', lastName: '', age: 0, id: 0, company: { title: '' } }))
                navigate('/edit')
            }}>Create New User</button>
            <div className='list-wrapper'>
                {
                    users.map((item, idx) => (
                        <UserWrapper key={`${item.id}-${idx}`} data={item} />
                    ))
                }
            </div>
        </div>
    )
}

export default UserList