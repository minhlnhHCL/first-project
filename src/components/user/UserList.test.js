import axios from "axios"
import reducer, { getUsers } from "../../store/userSlice"
import { renderWithRedux, screen } from "../../test-utils"
import { fetchData } from "../utils/fetchData"
import UserList from "./UserList"
import { useDispatch, useSelector } from "react-redux";
import { waitFor } from '@testing-library/react';

const sampleUser = [{
    id: 1,
    firstName: "Minh",
    lastName: "Luong",
    image: 'https://robohash.org/hicveldicta.png',
    age: 25,
    company: {
        title: "Front End Engineer",
    },
}, {
    id: 1,
    firstName: "Min",
    lastName: "Min",
    image: 'https://robohash.org/hicveldicta.png',
    age: 21,
    company: {
        title: "Front End Engineer",
    },
}]


// jest.mock('axios')

jest.mock("react-redux", () => ({
    ...jest.requireActual("react-redux"),
    useSelector: jest.fn(),
    useDispatch: jest.fn(),
}));



let getSpySuccess

describe('User List', () => {
    beforeAll(() => {
        jest.spyOn(axios, 'get').mockImplementation(() => {
            return new Promise((resolve, reject) => {
                return resolve({
                    data: {
                        users: [...sampleUser]
                    }
                });
            });
        });
        useDispatch.mockReturnValue(() => { });
        useSelector.mockReturnValue({
            users: []
        });

    })
    it('should run its side effect and be rendered', async () => {
        // expect(getSpySuccess).toHaveBeenCalledTimes(1);
        // expect(getSpySuccess).toBeCalledWith(
        //     'https://dummyjson.com/users'
        // );
        const { container } = renderWithRedux(<UserList />, {
            preloadedState: {
                users: [...sampleUser]
            }
        })
        // expect(useDispatch).toHaveBeenCalledTimes(1);
        expect(container).toMatchSnapshot();
        // const dispatch = jest.fn();
        // const getUsers = jest.fn()
        // useDispatch.mockReturnValue(dispatch);
        // await waitFor(() => expect(getUsers).toBeCalled());
        // expect(dispatch).toBeCalled()
        // expect(dispatch).toBeCalledWith(
        //     sampleUser
        // );

    })
})