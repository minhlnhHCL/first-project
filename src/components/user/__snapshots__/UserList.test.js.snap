// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User List should run its side effect and be rendered 1`] = `<div />`;
